Source/WebCore/page/scrolling/ScrollAnchoringController.cpp

```diff
@@ -25,30 +25,91 @@
 
 #include "config.h"
 #include "ScrollAnchoringController.h"
+#include "ElementIterator.h"
+#include "HTMLHtmlElement.h"
+#include "RenderBox.h"
+#include "TypedElementDescendantIteratorInlines.h"
+#include <wtf/text/TextStream.h>
 
 namespace WebCore {
 
 void ScrollAnchoringController::invalidateAnchorElement()
 {
-    m_anchorElement = nullptr;
-    m_lastPositionForAnchorElement = FloatPoint();
+    if (!m_midUpdatingScrollPositionForAnchorElement) {
+        m_anchorElement = nullptr;
+        m_lastOffsetForAnchorElement = { };
+        m_isQueuedForScrollPositionUpdate = false;
+        m_frameView.queueScrollableAreaForScrollAnchoringUpdate(m_frameView);
+    }
 }
 
-void ScrollAnchoringController::selectAnchorElement()
+ScrollOffset ScrollAnchoringController::computeOffset(RenderBox& candidate)
 {
-    // TODO: implement
-    // perform Anchor Element Selection algorithm on dom elements within frame view's viewable region
-    // Firefox appears to only iterate through frame tree under the current frame (unclear if
-    // looking through frame tree is sufficient)
-    // unclear to me where to call this from 
+    return ScrollOffset(FloatPoint(candidate.absoluteBoundingBoxRect().location() - m_frameView.layoutViewportRect().location()));
+}
+
+void ScrollAnchoringController::chooseAnchorElement()
+{
+    // TODO: fully implement anchor node selection algorithm in spec
+    // (specifically excluded subtrees and priority nodes)
+    // https://drafts.csswg.org/css-scroll-anchoring/#anchor-node-selection
+
+    auto* document = m_frameView.frame().document();
+    if (!document)
+        return;
+
+    Element* candidateAnchor = nullptr;
+    auto layoutViewport = m_frameView.layoutViewportRect();
+    for (auto& element : descendantsOfType<Element>(document->rootNode())) {
```

Comment:
This is will be wrong when this ScrollAnchoringController exists for an overflow scroll, because you're starting at the document root, and not the root of this scroller.

Reply:
Yeah this will be addressed in https://github.com/WebKit/WebKit/pull/17934. Added a todo for now.